### PR TITLE
Fix Tests in Arbitrum Branch

### PR DESCRIPTION
**Pull request by akshay111meher**
1. arb L1 gas calculation is mocked by a mocked contract.
2. benchmarks for existing contracts are fixed (but L1 gas consumed is again mocked)
3. Fix the possible tests

### CODE REVIEW

File: .openzeppelin/unknown-250.json
Comment (roshanrags) [inline .openzeppelin/unknown-250.json:1]: Why is this file removed?

File: package.json
Comment (roshanrags) [inline package.json:1]: hardhat has an "official" toolbox with what they consider recommended packages - https://hardhat.org/hardhat-runner/plugins/nomicfoundation-hardhat-toolbox

It has hardhat-chai-matchers over hardhat-waffle. It's nicer as well and seems to have better features like `revertedWithPanic`. What do we need waffle for?

File: contracts/mocks/ArbGasInfo.sol
Comment (roshanrags) [inline contracts/mocks/ArbGasInfo.sol:1]: We should probably use `deployMockContract` to deploy mocks instead of writing our own like this.

File: test/staking/RewardDelegators.ts
Comment (roshanrags) [inline test/staking/RewardDelegators.ts:1]: Isn't this file being replaced with RewardDelegatorsNew.ts

**Review by roshanrags — CHANGES_REQUESTED** [review]

File: package.json
Comment (akshay111meher) [reply package.json:1]: In benchmarks, the waffle package loads fixtures, but conflicts arise when combining waffle and chai-matchers due to conflicting definitions. Despite chai-matchers being better, waffle was chosen for prioritizing fixture loading in the benchmarks

Comment (roshanrags) [reply package.json:1]: hardhat-network-helpers has `loadFixture`, is that enough

File: contracts/mocks/ArbGasInfo.sol
Comment (akshay111meher) [reply contracts/mocks/ArbGasInfo.sol:1]: `deployMockContract` won't have typings generated for mock contract. Hence it is very painful to debug whenever some other contracts are changed. (through out the repo)

An intermediate step could be declare mock/ArbInfo.sol, deploy using `deployMockContract` and then cast to mock/ArbInfo.sol

File: test/staking/RewardDelegators.ts
Comment (akshay111meher) [reply test/staking/RewardDelegators.ts:1]: Both of files cover different tests. Will rename them to something proper..

File: .openzeppelin/unknown-250.json
Comment (akshay111meher) [reply .openzeppelin/unknown-250.json:1]: chain 250 corresponds to Fantom and afaik, this repo has no interactions with Fantom

Comment (roshanrags) [reply .openzeppelin/unknown-250.json:1]: we did deploy some stuff there, leave the file as it is

File: package.json
Comment (akshay111meher) [reply package.json:1]: reverted back to original

File: .openzeppelin/unknown-250.json
Comment (akshay111meher) [reply .openzeppelin/unknown-250.json:1]: Added back

Comment (roshanrags) [reply .openzeppelin/unknown-250.json:1]: it's still deleted :sweat_smile:

File: hardhat.config.ts [hunk 1]
```diff
@@ -4,7 +4,8 @@ import "@nomiclabs/hardhat-etherscan";
 import '@openzeppelin/hardhat-upgrades';
 import 'solidity-coverage';
 import "hardhat-gas-reporter";
-import '@nomicfoundation/hardhat-chai-matchers';
+import "@nomicfoundation/hardhat-chai-matchers";
+import "@nomicfoundation/hardhat-network-helpers";
```
Comment (roshanrags) [inline hardhat.config.ts]: don't think this is needed, we import it directly in the test files anyway

File: benchmarks/ClusterRewards.ts [hunk 2]
```diff
@@ -1,8 +1,11 @@
-import { ethers, waffle } from "hardhat";
+import { ethers } from "hardhat";
+import { loadFixture } from '@nomicfoundation/hardhat-network-helpers'
 import { benchmark as benchmarkDeployment } from "./helpers/deployment";
 import { initDataFixture } from "./fixtures/ClusterRewards";
 import { BigNumber, BigNumberish, constants, Contract, PopulatedTransaction, Signer, utils } from "ethers";
 import { randomlyDivideInXPieces, skipTime } from "./helpers/util";
+import { ClusterRewards, ClusterRewards__factory } from "../typechain-types";
+import { IArbGasInfo__factory } from "../typechain-types/factories/contracts/staking/ClusterSelector.sol";
```
Comment (roshanr95) [inline benchmarks/ClusterRewards.ts:8]: imports ClusterSelector instead of ArbGasInfo

File: test/staking/ClusterRewards.ts [hunk 3]
```diff
@@ -912,7 +912,7 @@ describe("ClusterRewards submit tickets", function() {
     await time.increaseTo(startTime + 34 * 86400);
 
     // reverted because of overflow in args
-    await expect(clusterRewards.connect(signers[5])["issueTickets(bytes32,uint24,uint16[])"](ETHHASH, epochWithRewards, [MAX_TICKETS_1_pc.mul(10), MAX_TICKETS_1_pc.mul(20), MAX_TICKETS_1_pc.mul(30), MAX_TICKETS.add(1)])).to.be.revertedWith("");
+    await expect(clusterRewards.connect(signers[5])["issueTickets(bytes32,uint24,uint16[])"](ETHHASH, epochWithRewards, [MAX_TICKETS_1_pc.mul(10), MAX_TICKETS_1_pc.mul(20), MAX_TICKETS_1_pc.mul(30), MAX_TICKETS_1_pc.mul(40).add(10000)])).to.be.revertedWith("CRW:IPRT-Total ticket count invalid");
```
Comment (roshanrags) [inline test/staking/ClusterRewards.ts]: is 10000 the minimum I need to add to get it to overflow?

File: test/staking/ClusterSelector.ts [hunk 4]
```diff
@@ -1374,6 +1374,7 @@ describe("ClusterSelector", function() {
 
   it("current epoch", async () => {
     const epochLength = parseInt((await clusterSelector.EPOCH_LENGTH()).toString());
+    // Todo: check if the revert statement is sufficient
```
Comment (roshanrags) [inline test/staking/ClusterSelector.ts]: resolve before merge

File: test/staking/ClusterSelector.ts [hunk 5]
```diff
@@ -1472,6 +1473,7 @@ describe("ClusterSelector", function() {
 
     await expect(clusterSelector.updateMissingClusters(2)).to.be.revertedWith("cannot update future epochs");
     await clusterSelector.updateMissingClusters(1);
+    // Todo: check if the revert statement is sufficient
```
Comment (roshanrags) [inline test/staking/ClusterSelector.ts]: resolve before merge

File: test/staking/Integration.ts [hunk 6]
```diff
@@ -187,18 +188,21 @@ describe("Integration", function () {
...
         [
           await clusterSelectorAdmin.getAddress(),
-          rewardDelegators.address,
-          BN.from(10).pow(20).toString(),
+          rewardDelegators.address
         ],
         {
           kind: "uups",
-          constructorArgs: [await receiverStaking.START_TIME(), await receiverStaking.EPOCH_LENGTH()],
+          // Todo: check the max value rewards and gas refund
```
Comment (roshanrags) [inline test/staking/Integration.ts]: resolve before merge

File: test/staking/Integration.ts [hunk 7]
```diff
@@ -456,26 +472,30 @@ describe("Integration", function () {
         mpondTokenId
       );
 
-      // console.log(pondRewards.map((a) => a.toString()));
-      // console.log(mpondRewards.map((a) => a.toString()));
+      console.log(pondRewards.map((a) => a.toString()));
+      console.log(mpondRewards.map((a) => a.toString()));
```
Comment (roshanrags) [inline test/staking/Integration.ts]: comment out, don't print in the middle of tests

File: benchmarks/ClusterRewards.ts (hunk 2 above)
Comment (roshanrags) [inline benchmarks/ClusterRewards.ts]: can't IArbGasInfo__factory be imported from "../typechain-types" as well?

Comment (prateekreddy) [reply benchmarks/ClusterRewards.ts]: This seems to be unused, so can be removed ?

File: benchmarks/ClusterRewards.ts [hunk 8]
```diff
@@ -28,7 +31,7 @@ describe("Cluster Rewards", async () => {
         let pond: Contract;
         let receiverStaking: Contract;
         let clusterSelector: Contract;
```
Comment (prateekreddy) [inline benchmarks/ClusterRewards.ts]: Should we replace all contracts with the actual contract types ?

File: test/staking/ClusterRewards.ts (hunk 3 above)
Comment (akshay111meher) [reply test/staking/ClusterRewards.ts]: no, 
it was added because some values got floored due to division. 
Changed it now

File: benchmarks/ClusterRewards.ts [hunk 9]
```diff
@@ -49,52 +52,56 @@ describe("Cluster Rewards", async () => {
...
+                clusterRewards: tempClusterRewards,
                 admin,
                 rewardDelegatorsMock,
                 nodesInserted,
                 receivers,
                 receiverSigners
-            } = await waffle.loadFixture(initDataFixture));
+            } = await loadFixture(initDataFixture));
 
+            clusterRewards = ClusterRewards__factory.connect(tempClusterRewards.address, admin)
```
Comment (prateekreddy) [inline benchmarks/ClusterRewards.ts]: Can we move this to `initDataFixture` so that the return type is `ClusterRewards` instead of `Contract`

File: benchmarks/ClusterRewards.ts [hunk 10]
```diff
@@ -49,52 +52,56 @@ describe("Cluster Rewards", async () => {
...
-                await skipTime(EPOCH_LENGTH);
+            console.log(tickets.map(a => a.toString()))
+            // skip to next epoch so that tickets can be distributed for previous epoch
+            await skipTime(EPOCH_LENGTH);
 
-                const gasEstimate = await clusterRewards.connect(selectedReceiverSigner)["issueTickets(bytes32,uint24,uint16[])"].estimateGas(
-                    ethers.utils.id("ETH"), epoch, tickets
-                );
-                console.log(`gas used for ${i} cluster : ${gasEstimate.toNumber()}`);
+            const gasEstimate = await clusterRewards.connect(selectedReceiverSigner).estimateGas["issueTickets(bytes32,uint24,uint16[])"](
```
Comment (prateekreddy) [inline benchmarks/ClusterRewards.ts]: This function was supposed to calculate gas if we submit tickets for 1-5 clusters, I think the updated version only calculates for 5 clusters. Any specific reason for doing this ?

File: benchmarks/ClusterRewards.ts [hunk 11]
```diff
@@ -49,52 +52,56 @@ describe("Cluster Rewards", async () => {
...
+            await skipTime(EPOCH_LENGTH);
 
-                let epoch = await clusterSelector.getCurrentEpoch();
-                const tickets: BigNumber[] = randomlyDivideInXPieces(MAX_TICKETS, i);
+            let epoch = await clusterSelector.getCurrentEpoch();
+            const tickets: BigNumber[] = randomlyDivideInXPieces(MAX_TICKETS, 5);
 
-                // skip to next epoch so that tickets can be distributed for previous epoch
-                await skipTime(EPOCH_LENGTH);
+            console.log(tickets.map(a => a.toString()))
```
Comment (prateekreddy) [inline benchmarks/ClusterRewards.ts]: Remove this ?

File: benchmarks/ClusterRewards.ts [hunk 12]
```diff
@@ -49,52 +52,56 @@ describe("Cluster Rewards", async () => {
...
+            await skipTime(EPOCH_LENGTH);
 
-                const gasEstimate = await clusterRewards.connect(selectedReceiverSigner)["issueTickets(bytes32,uint24,uint16[])"].estimateGas(
-                    ethers.utils.id("ETH"), epoch, tickets
-                );
-                console.log(`gas used for ${i} cluster : ${gasEstimate.toNumber()}`);
+            const gasEstimate = await clusterRewards.connect(selectedReceiverSigner).estimateGas["issueTickets(bytes32,uint24,uint16[])"](
+                ethers.utils.id("ETH"), epoch, tickets
+            );
+            console.log(`gas used for ${5} cluster : ${gasEstimate.toNumber()}`);
```
Comment (prateekreddy) [inline benchmarks/ClusterRewards.ts]: This gas estimate is exclusively L2 gas but we want to know L1 gas as well, that part seems to be missing

File: benchmarks/ClusterRewards.ts [hunk 13]
```diff
@@ -137,16 +144,18 @@ describe("Cluster Rewards", async () => {
...
             }
 
-            const tx = await clusterRewards["issueTickets(bytes32,uint24,(uint16[],uint8,bytes32,bytes32)[])"](
-                ethers.utils.id("ETH"), epoch, signedTickets
-            );
 
-            const receipt = await tx.wait();
-            console.log(`gas used : ${receipt.gasUsed.sub(21000).toNumber()}`);
-            console.log(receipt.gasUsed.mul(1600).mul(6).mul(365).div(BigNumber.from(10).pow(10)).toString());
+            // TODO: this function definitation has been removed. Confirm whether to remove it from benchmarks
```
Comment (prateekreddy) [inline benchmarks/ClusterRewards.ts]: Any reason this benchmark is commented out ?

File: benchmarks/ClusterRewards.ts [hunk 14]
```diff
@@ -194,7 +203,7 @@ describe("Cluster Rewards", async () => {
             console.log(gasEstimate.add(l1GasInL2).mul(1600).mul(50).mul(365).div(BigNumber.from(10).pow(10)).toString());
```
Comment (prateekreddy) [inline benchmarks/ClusterRewards.ts]: We can probably abstract this into a new functions which calculates cost per year stuff and consoles can be improved.

File: benchmarks/ClusterSelector.ts [hunk 15]
```diff
@@ -35,7 +36,7 @@ describe("Cluster Rewards", async () => {
                 admin,
                 rewardDelegatorsMock,
                 nodesInserted,
-            } = await waffle.loadFixture(initDataFixture));
+            } = await loadFixture(initDataFixture));
```
Comment (prateekreddy) [inline benchmarks/ClusterSelector.ts:44]: Is there a difference between using waffle loadFixture vs loadFixture from hardhat-network-helpers ?

File: benchmarks/fixtures/ClusterRewards.ts [hunk 16]
```diff
@@ -18,14 +21,22 @@ export async function deployFixture() {
...
+    // mockArbGas.setPrices(10000, 10000, 10000);
 
+    const receiverStaking = await deployReceiverStaking(addrs[0], blockData.timestamp, EPOCH_LENGTH, pond.address, true);
+    const clusterSelector = await deployClusterSelector("ETH",
+    addrs[1],
+    arbGasContract,
+    addrs[0],
+    blockData.timestamp,
+    EPOCH_LENGTH,
+    "10000000",
```
Comment (prateekreddy) [inline benchmarks/fixtures/ClusterRewards.ts]: Can we declare this as well as constant probably ?

File: benchmarks/fixtures/ClusterRewards.ts [hunk 17]
```diff
@@ -18,14 +21,22 @@ export async function deployFixture() {
         kind: "uups",
     });
 
-    const receiverStaking = await deployReceiverStaking(addrs[0], blockData.timestamp, EPOCH_LENGTH, pond.address, true);
-
-    const clusterSelector = await deployClusterSelector("ETH", addrs[1], "0x000000000000000000000000000000000000006C", addrs[0], blockData.timestamp, EPOCH_LENGTH, ethers.utils.parseEther('1').toString(), true);
+    // const mockArbGas = await new ArbGasInfo__factory().connect(signers[0]).deploy();
```
Comment (prateekreddy) [inline benchmarks/fixtures/ClusterRewards.ts]: remove these comments and above commented stuff as well

File: test/staking/Integration.ts [hunk 18]
```diff
@@ -292,6 +298,15 @@ describe("Integration", function () {
     const delegatorsToUse = 20;
 
     saveAndRestoreStateToParent(async () => {
+      // skip epoch till switching
+      let currentEpoch = (await clusterSelectors[0].getCurrentEpoch()).toNumber();
+      const switchingEpoch = BN.from(33).mul(86400).div(4 * 3600).toNumber() // coded as private variables in contract
```
Comment (prateekreddy) [inline test/staking/Integration.ts:310]: We can just remove switching stuff right ? why was this needed ?

File: test/staking/Integration.ts [hunk 19]
```diff
@@ -376,19 +392,19 @@ describe("Integration", function () {
       for (let index = 0; index < pondRewards.length - 1; index++) {
         const element = pondRewards[index];
         expect(element).gt(0);
-        expect(element).to.be.eq(pondRewards[index + 1]);
+        expect(element).to.be.closeTo(pondRewards[index + 1], pondRewards[index + 1].div(1000));
```
Comment (prateekreddy) [inline test/staking/Integration.ts]: Is 0.1% the least deviation we can add ?

File: test/staking/Integration.ts [hunk 20]
```diff
@@ -344,13 +359,14 @@ describe("Integration", function () {
         const cluster = clusterAddresses[index];
         expect(clusters.includes(cluster)).to.be.false;
       }
-    });
+    }); 
 
     it("Only 1 receiver, all cluster should get equal reward", async () => {
       const receiver = receivers[0];
-      const totalWeight = BN.from(10).pow(18);
+      const totalWeight = BN.from(2).pow(16);
```
Comment (prateekreddy) [inline test/staking/Integration.ts]: This should be 2^16 - 1 instead

File: test/staking/Integration.ts (hunk 6 above)
Comment (akshay111meher) [reply test/staking/Integration.ts]: done.

File: test/staking/Integration.ts [hunk 21]
```diff
@@ -456,26 +472,30 @@ describe("Integration", function () {
...
-        expect(element).eq(0);
-      }
-      expect(pondRewards[pondRewards.length - 1]).gt(0);
+      let pondZeroRewardClusterCount: number = 0
+      pondRewards.forEach(a => {
+        if(a.gt(0)){
+          pondZeroRewardClusterCount++
+        }
+      })
+      expect(pondZeroRewardClusterCount).eq(4)
```
Comment (prateekreddy) [inline test/staking/Integration.ts]: This leaves open the possibility that some other cluster is getting the rewards instead of the one supposed to.

File: test/staking/Integration.ts [hunk 22]
```diff
@@ -456,26 +472,30 @@ describe("Integration", function () {
...
-        expect(element).eq(0);
-      }
-      expect(mpondRewards[mpondRewards.length - 1]).gt(0);
+      let mpondZeroRewardClusterCount: number = 0
+      mpondRewards.forEach(a => {
+        if(a.gt(0)){
+          mpondZeroRewardClusterCount++
+        }
+      })
+      expect(mpondZeroRewardClusterCount).eq(4)
```
Comment (prateekreddy) [inline test/staking/Integration.ts]: This leaves open the possibility that some other cluster is getting the rewards instead of the one supposed to.

File: test/staking/Integration.ts (hunk 7 above)
Comment (akshay111meher) [reply test/staking/Integration.ts]: commented

File: test/staking/Integration.ts [hunk 23]
```diff
@@ -456,26 +472,30 @@ describe("Integration", function () {
...
     });
 
     it("1 receiver, ticket ratio 1:1:1:1:1, all delegator get equal rewards, totalReward < rewardPerEpoch", async () => {
       const receiver = receivers[0];
-      let equiWeightedTickets = BN.from(10).pow(18).div(totalClusters);
-      const weights = [equiWeightedTickets, equiWeightedTickets, equiWeightedTickets, equiWeightedTickets, equiWeightedTickets];
+      let equiWeightedTickets = BN.from(2).pow(16).div(totalClusters);
```
Comment (prateekreddy) [inline test/staking/Integration.ts]: use 2^16 - 1 instead

File: benchmarks/ClusterRewards.ts (hunk 2 above)
Comment (akshay111meher) [reply benchmarks/ClusterRewards.ts]: Import not used, hence removed now.

File: hardhat.config.ts (hunk 1 above)
Comment (akshay111meher) [reply hardhat.config.ts]: when commented out it this happens

`Invalid Chai property: revertedWith`

File: test/staking/Integration.ts [hunk 24]
```diff
@@ -505,7 +525,10 @@ describe("Integration", function () {
       );
       // console.log(delegatorRewardsRecevied.map(a => a.toString()));
       expect(delegatorRewardsRecevied.every((a) => a.gt(0))).to.be.true;
-      expect(delegatorRewardsRecevied.every((a) => a.eq(delegatorRewardsRecevied[0]))).to.be.true;
+      for (let index = 0; index < delegatorRewardsRecevied.length - 1; index++) {
+        const element = delegatorRewardsRecevied[index];
+        expect(element).to.be.closeTo(delegatorRewardsRecevied[index + 1], delegatorRewardsRecevied[index + 1].div(1000))
```
Comment (prateekreddy) [inline test/staking/Integration.ts]: Why close to, it should be based on whatever tickets were given to each na, sure probably not everyone has equal tickets, then the distribution should be based on ticket ratios right ?

File: test/staking/Integration.ts [hunk 25]
```diff
@@ -514,7 +537,7 @@ describe("Integration", function () {
       const totalExpectedRewardToBeDistributed = REWARD_PER_EPOCH.mul(supportedNetworksWeights[0]).div(100);
 
       expect(totalPondDistributed).lt(totalExpectedRewardToBeDistributed);
-      expect(totalExpectedRewardToBeDistributed).is.closeTo(totalPondDistributed, 10);
+      expect(totalExpectedRewardToBeDistributed).is.closeTo(totalPondDistributed, totalPondDistributed.div(10000));
```
Comment (prateekreddy) [inline test/staking/Integration.ts:606]: The common question across all tests is if the error is as high as 0.1%

File: test/staking/Integration.ts [hunk 26]
```diff
@@ -523,7 +546,7 @@ describe("Integration", function () {
 
       await receiverDeposit(pond, receiverStaking, receiver1, minPondToUseByReceiver.div(2));
       await receiverDeposit(pond, receiverStaking, receiver2, minPondToUseByReceiver);
-      const fraction = BN.from(10).pow(18);
+      const fraction = BN.from(2).pow(16).sub(1);
```
Comment (prateekreddy) [inline test/staking/Integration.ts]: Probably declare a constant somewhere, easier that changing everywhere na

File: test/staking/Integration.ts [hunk 27]
```diff
@@ -626,25 +649,22 @@ describe("Integration", function () {
       let currentEpoch = (await clusterSelectors[0].getCurrentEpoch()).toNumber();
       let currentPlusOne = BN.from(currentEpoch).add(1).toString();
       await clusterSelectors[0].selectClusters();
+
+      // skipping only for issuing the rewards
       await skipEpoch();
+      await skipEpoch();
+      await skipEpoch();
```
Comment (prateekreddy) [inline test/staking/Integration.ts:719]: What is the reason for skipping 3 epochs ?

File: test/staking/Integration.ts [hunk 28]
```diff
@@ -687,6 +707,15 @@ describe("Integration", function () {
     let mpondDelegatedByMaticDelegators: string[] = [];
 
     saveAndRestoreStateToParent(async () => {
+      // skip epoch till switching
+      let currentEpoch = (await clusterSelectors[0].getCurrentEpoch()).toNumber();
+      const switchingEpoch = BN.from(33).mul(86400).div(4 * 3600).toNumber() // coded as private variables in contract
```
Comment (prateekreddy) [inline test/staking/Integration.ts]: what does this switching thing do ?

File: test/staking/Integration.ts [hunk 29]
```diff
@@ -873,7 +902,7 @@ describe("Integration", function () {
     ): Promise<void> => {
       const epoch = BN.from(clusterSelectedAtEpoch).sub(1).toString();
       const weights: BN[] = [];
-      let pendingFraction = BN.from(10).pow(18);
+      let pendingFraction = BN.from(2).pow(16);
```
Comment (prateekreddy) [inline test/staking/Integration.ts]: Use 2^16 - 1 instead

File: test/staking/Integration.ts [hunk 30]
```diff
@@ -1061,17 +1090,17 @@ const issueTicketsForClusters = async (
...
-
-  // console.log({
-  //   currentPlusOne,
-  //   networkIds,
-  //   weights: weights.map((a) => a.toString()),
-  //   clustersToIssueTicketsTo,
-  //   selectedClusters: await clusterSelectorInstance.getClusters(currentPlusOne),
-  // });
+  
+  console.log({
```
Comment (prateekreddy) [inline test/staking/Integration.ts]: Remove logs from tests

File: test/staking/Integration.ts [hunk 31]
```diff
@@ -1089,10 +1118,10 @@ const issueTicketsForClusters = async (
 
   // weights array has to re-arranged in order of selected clusters
   const new_weights = new_order_of_weights(weights, clustersToIssueTicketsTo, await clusterSelectorInstance.getClusters(currentPlusOne));
-
+  console.log(new_weights)
```
Comment (prateekreddy) [inline test/staking/Integration.ts]: Remove logs from tests

File: benchmarks/ClusterRewards.ts (hunk 8 above)
Comment (akshay111meher) [reply benchmarks/ClusterRewards.ts]: Done. Expect for the mocks

File: benchmarks/ClusterRewards.ts (hunk 9 above)
Comment (akshay111meher) [reply benchmarks/ClusterRewards.ts]: moved

File: benchmarks/ClusterRewards.ts (hunk 10 above)
Comment (akshay111meher) [reply benchmarks/ClusterRewards.ts]: Recollect that it was reverting when input was 1 to 4, but not 5.
Reverted to original. 
To debug

File: benchmarks/ClusterRewards.ts (hunk 11 above)
Comment (akshay111meher) [reply benchmarks/ClusterRewards.ts]: reverted to original

File: benchmarks/ClusterRewards.ts (hunk 12 above)
Comment (akshay111meher) [reply benchmarks/ClusterRewards.ts]: reverted.

File: benchmarks/ClusterRewards.ts (hunk 13 above)
Comment (akshay111meher) [reply benchmarks/ClusterRewards.ts]: There is no function that matches `issueTickets(bytes32,uint24,(uint16[],uint8,bytes32,bytes32)[])` in the contract. (even in base branch)

If correct, then shall remove the whole test

File: benchmarks/ClusterSelector.ts (hunk 15 above)
Comment (akshay111meher) [reply benchmarks/ClusterSelector.ts:44]: waffle is inducing some other un-necessay things into the repo...
hence using other package to loadFixtures.

Both are identical wrt fixtures

File: benchmarks/fixtures/ClusterRewards.ts (hunk 16 above)
Comment (akshay111meher) [reply benchmarks/fixtures/ClusterRewards.ts]: done

File: benchmarks/fixtures/ClusterRewards.ts (hunk 17 above)
Comment (akshay111meher) [reply benchmarks/fixtures/ClusterRewards.ts]: removed

File: test/staking/Integration.ts (hunk 18 above)
Comment (akshay111meher) [reply test/staking/Integration.ts:310]: will be removed in upcoming commit. as of this branch switching is not removed... and tests in preview of arbitrum branch

File: test/staking/Integration.ts (hunk 19 above)
Comment (akshay111meher) [reply test/staking/Integration.ts]: due to nature of total tickets (2^16),

The deviations could be of order `1/10000`
Updated the deviation to `1/10000`

File: test/staking/Integration.ts (hunk 20 above)
Comment (akshay111meher) [reply test/staking/Integration.ts]: changed

File: test/staking/Integration.ts (hunk 21 above)
Comment (akshay111meher) [reply test/staking/Integration.ts]: In that specific test, it is ensured only a given set of clusters are selected

File: test/staking/Integration.ts (hunk 22 above)
Comment (akshay111meher) [reply test/staking/Integration.ts]: In that specific test, it is ensured only a given set of clusters are selected.

File: test/staking/Integration.ts (hunk 23 above)
Comment (akshay111meher) [reply test/staking/Integration.ts]: Done

File: test/staking/Integration.ts (hunk 24 above)
Comment (akshay111meher) [reply test/staking/Integration.ts]: In the given test_scenario, every everyone has equal tickets.

File: test/staking/Integration.ts (hunk 25 above)
Comment (akshay111meher) [reply test/staking/Integration.ts:606]: max tickets are 65536   .......  ( minus 1) 
The error rate will be in the multiples of `1/65536`

when tickets are distributed equally i,e `[13072, 13072, 13072, 13072, 13073] (one of them will be ±1)`
the error rate increases to `1/13072`

Hence, in after running the tests multiple times, `1/10000` error rate was a chosen

File: test/staking/Integration.ts (hunk 26 above)
Comment (akshay111meher) [reply test/staking/Integration.ts]: done

File: test/staking/Integration.ts (hunk 27 above)
Comment (akshay111meher) [reply test/staking/Integration.ts:719]: In most cases where `skipEpoch` is called multiple times, we wan't to go far ahead in time. It could be any number more than that..

File: test/staking/Integration.ts (hunk 28 above)
Comment (akshay111meher) [reply test/staking/Integration.ts]: moves ahead in time if there switching period is enabled.

File: test/staking/Integration.ts (hunk 29 above)
Comment (akshay111meher) [reply test/staking/Integration.ts]: done

File: test/staking/Integration.ts (hunk 30 above)
Comment (akshay111meher) [reply test/staking/Integration.ts]: commented

File: test/staking/Integration.ts (hunk 31 above)
Comment (akshay111meher) [reply test/staking/Integration.ts]: removed

File: benchmarks/ClusterRewards.ts (hunk 10 above)
Comment (prateekreddy) [reply benchmarks/ClusterRewards.ts]: Ahh.. right, you have to submit 0s for rest of the clusters ig. Because clusters selected is fixed to 5 now.

File: benchmarks/ClusterRewards.ts (hunk 13 above)
Comment (prateekreddy) [reply benchmarks/ClusterRewards.ts]: got it, we can remove this benchmark in that case.

File: hardhat.config.ts (hunk 1 above)
Comment (roshanrags) [reply hardhat.config.ts]: I'm referring to hardhat-network-helpers, not hardhat-chai-matchers

**Review by prateekreddy — APPROVED** [review]

**Review by roshanrags — APPROVED** [review]
